cmd/kubectl-gadget/tcptop.go

```diff
@@ -0,0 +1,314 @@
+// Copyright 2019-2021 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"strconv"
+	"strings"
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/spf13/cobra"
+	"golang.org/x/term"
+
+	"github.com/kinvolk/inspektor-gadget/cmd/kubectl-gadget/utils"
+	"github.com/kinvolk/inspektor-gadget/pkg/gadgets/tcptop/types"
+)
+
+var (
+	nodeTCPStats map[string][]types.Stats
+	mtx          sync.Mutex
+	ended        bool = false
+)
+
+var (
+	// arguments
+	intervalOutput int = 1
+
+	// flags
+	maximumRows int
+	sortTypeStr string
+	sortType    types.SortBy
+	filteredPid uint
+	family      uint
+)
+
+func translateFamily(family uint) (string, error) {
+	switch family {
+	case 4:
+		return strconv.Itoa(syscall.AF_INET), nil
+	case 6:
+		return strconv.Itoa(syscall.AF_INET6), nil
+	case 0:
+		return "-1", nil
+	default:
+		return "", fmt.Errorf("IP version is either 4 or 6, %d was given", family)
+	}
+}
+
+var tcptopCmd = &cobra.Command{
+	Use:   "tcptop [interval]",
+	Short: "Trace reads and writes by file, with container details",
+	RunE: func(cmd *cobra.Command, args []string) error {
+		var err error
+
+		nodeTCPStats = make(map[string][]types.Stats)
+
+		if len(args) == 1 {
+			intervalOutput, err = strconv.Atoi(args[0])
+			if err != nil {
+				return fmt.Errorf("%q is not a valid value for intervalOutput", args[0])
+			}
+		}
+
+		familyStr, err := translateFamily(family)
+		if err != nil {
+			return err
+		}
+
+		pidStr := "-1"
+		if filteredPid != 0 {
+			pidStr = strconv.FormatUint(uint64(filteredPid), 10)
+		}
+
+		config := &utils.TraceConfig{
+			GadgetName:       "tcptop",
+			Operation:        "start",
+			TraceOutputMode:  "Stream",
+			TraceOutputState: "Started",
+			CommonFlags:      &params,
+			Parameters: map[string]string{
+				"max_rows":       strconv.Itoa(maximumRows),
+				"intervalOutput": strconv.Itoa(intervalOutput),
+				"sort_by":        sortTypeStr,
+				"pid":            pidStr,
+				"family":         familyStr,
+			},
+		}
+
+		startoutput()
+
+		if err := utils.RunTraceStreamCallback(config, hook); err != nil {
+			return fmt.Errorf("error running trace: %w", err)
+		}
+
+		return nil
+	},
+	SilenceUsage: true,
+	PreRunE: func(cmd *cobra.Command, args []string) error {
+		var err error
+		sortType, err = types.ParseSortBy(sortTypeStr)
+		if err != nil {
+			return err
+		}
```

Comment:
I did not understand.

Reply:
I mean, why not do all the parsing in `PreRunE` instead of doing it for `sort` in `PreRunE` and for the args (`interval`), `family` and `pid` in `RunE`?

Did I explain myself better?

Reply:
Thank you, it is clearer now.
I think I cannot in the way I handle the thing now, I will push it soon.